### ISSUE TITLE
Switching AI model: Tensorflow + Basic Information on How Machine Learning Works (Required if using TensorFlow) 03.29

### ISSUE THREAD

**Issue by p-thotapalli** [upparao4860-sketch/Uppara-PPP-2025#80]
**Reason for Switching from Pytorch to Tensorflow**

It's actually pretty simple. I don't have a GPU (before I was relying on Om's computer), and TensorFlow allows people to use its model through CPU power or even on mobile devices. 

Other models also do not need the use of a GPU, but TensorFlow is a really popular AI model that people use for machine learning, so there are many, many more videos on how to use TensorFlow and different ways it can be used.

TensorFlow is also just simple. Training a model in a short amount of time with PyTorch is extremely hard and requires a huge amount of code. TensorFlow is integrated with the Keras API, which allows the user to not require large amounts of code to train a model (if it isn't that intensive, you could just use .train with TensorFlow).

**What is a Neural Network?**

This section is just a simple refresher, so it won't be as long as the other sections.

Video 1: [Link](https://www.youtube.com/watch?v=aircAruvnKk) 
(This is by Three Blue One Brown and looks at neural networks from a mathematical perspective, which is pretty interesting)

_This video mainly talks only about multi-layered neural networks, but this is what I plan on using for this project anyways._

**Neurons**

The neurons on a neural network, in the most simplistic form holds a number between 0 and 1. With image recognition, each neuron would be a pixel on the image. 

The example the video uses is a handwritten number on a 28x28 image. Each neuron would take up a pixel and hold a number. In this instance, it is the gray-scale value of each pixel (Maybe this might be a similar process to how Mediapipe works? Of course, not gray-scale, but finding the actual points on the body).

<img width="828" height="733" alt="Image" src="https://github.com/user-attachments/assets/07e31241-9bf8-49d7-9760-d18a2e1cf59e" />

The actual number inside a neuron is called Activation.

**Layers**

I'm going to continue to use the video's example. The first layer is going to be each of the individual pixels (28x28, so 784 neurons on the first layer).

The last layer would only be 10 neurons, 0-9. The activation number inside those neurons would be how much the system thinks that the number given as the input is. 

The layers between the first and last layer are called "Hidden Layers". They can be set to any number of neurons, and they would mean different things. Activations in one layer determine the activations in the next layer.

Video 2: [Link](https://www.youtube.com/watch?v=IHZwWFHWa-w)

**Gradient Descent**

Uh, learning this took me around 40 minutes, but the basic idea is that when testing the data, we do a certain input. Then, we compare the machine's output to the test data to see the "cost"/how far off the data is. By doing this repeatedly, the machine will slowly move toward a local minimum. By finding this local minimum, the machine will start to produce more accurate results. 

It's kinda interesting to see the calculus I learned in class being used in coding. However, I am not learning multivariable calculus ahead of time.

**Comment by drewmca-pkwy** [comment]
@p-thotapalli Puneeth, this looks great. I like the amount of detail and the connections to other areas of your studies outside of Spark!. I'm confused about the number minutes. In the body of the post you mention working on this for 40 minutes (that's about the runtime of the videos you linked), and the estimate is 210 minutes. Can you elaborate on the difference between the two? I could see about 60 - 70 minutes for the recording of your thoughts and additional questions. I'll need more info to understand 210 minutes. 

When you add that in a comment below, I'll update IC. Till then, we'll mark this one missing so we don't forget to come back to it.

**Comment by p-thotapalli** [comment]
There was another video I was watching/following on TensorFlow and using it to predict survivors on the Titanic. This was what I was doing for a good amount of time before I started to realize at the 2-3 hour mark of the video that I kinda had no idea on the basics of how machine learning and neural networks actually worked. With TensorFlow being a website that need more of a deeper understanding than PyTorch, I decided to stop watching that video and going back to the basics. I now realized I probably should have put this information at the top of the card.

This was the image that made me turn back:

<img width="1073" height="730" alt="Image" src="https://github.com/user-attachments/assets/fb3b7818-8b4b-4f3a-94f4-c088149266ac" />

This is the video (He did explain Neural Networks at the begining but I couldn't really understand it): [Link](https://www.youtube.com/watch?v=tPYj3fFJGjk&t=11168s)
